File: Problems/1143.Longest-Common-Subsequence/LongestCommonSubsequence.go
package main

import "fmt"

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

// Runtime: 0 ms, faster than 100.00% of Go online submissions for Longest Common Subsequence.
func longestCommonSubsequence(text1 string, text2 string) int {
	len1 := len(text1)
	len2 := len(text2)
	dp := make([][]int, len1+1)
	for i := range dp {
		dp[i] = make([]int, len2+1)
	}
	for r := 1; r <= len1; r++ {
		for c := 1; c <= len2; c++ {
			if text1[r-1] == text2[c-1] {
				dp[r][c] = dp[r-1][c-1] + 1
			} else {
				dp[r][c] = max(dp[r-1][c], dp[r][c-1])
			}
		}
	}

	return dp[len1][len2]
}

func main() {
	fmt.Println(longestCommonSubsequence("babcde", "ace"))
}

File: leetcode/1143.Longest-Common-Subsequence/LongestCommonSubsequence.py
class Solution:

    # Runtime: 376 ms, faster than 90.48% of Python3 online submissions for Longest Common Subsequence.
    def longestCommonSubsequence(self, text1: str, text2: str) -> int:
        len1 = len(text1)
        len2 = len(text2)
        dp = [[0]*(len2+1) for _ in range(len1+1)]
        
        for r in range(1, len1+1):
            for c in range(1, len2+1):
                if text1[r - 1] == text2[c-1]:
                    dp[r][c] = dp[r-1][c-1] + 1
                else:
                    dp[r][c] = max(dp[r-1][c], dp[r][c-1])

        # from pprint import pprint
        # pprint(dp)
        return dp[len1][len2]

    # Runtime: 336 ms, faster than 96.56% of Python3 online submissions for Longest Common Subsequence.
    def longestCommonSubsequenceV2(self, text1: str, text2: str) -> int:
        len1 = len(text1)
        len2 = len(text2)
        dp = [[0]*(len2+1) for _ in range(len1+1)]
        
        for r in range(1, len1+1):
            for c in range(1, len2+1):
                if text1[r - 1] == text2[c-1]:
                    dp[r][c] = dp[r-1][c-1] + 1
                else:
                    dp[r][c] = dp[r-1][c] if dp[r-1][c] > dp[r][c-1] else dp[r][c-1] 
        return dp[len1][len2]

if __name__ == "__main__":
    Solution().longestCommonSubsequence("babcde", "acxe")
    Solution().longestCommonSubsequence("babccde", "acxe")
    Solution().longestCommonSubsequence("abcba", "abcbcbaaaaaa")